src/radical/pilot/agent/executing/base.py

```diff
@@ -375,22 +375,19 @@ def _extend_pre_exec(self, td, ranks=None):
         # FIXME: need to distinguish between logical and physical IDs
 
         if td['threading_type'] == rpc.OpenMP:
-            # for future updates: if task ranks are heterogeneous in terms of
+            # for future updates: if task slots are heterogeneous in terms of
             #                     number of threads, then the following string
             #                     should be converted into dictionary (per rank)
             num_threads = td.get('cores_per_rank', 1)
             td['pre_exec'].append('export OMP_NUM_THREADS=%d' % num_threads)
 
-        if td['gpus_per_rank'] and td['gpu_type'] == rpc.CUDA and ranks:
+        if td['gpus_per_rank'] and td['gpu_type'] == rpc.CUDA and slots:
             # equivalent to the 'physical' value for original `cvd_id_mode`
-            rank_id  = 0
             rank_env = {}
-            for slot_ranks in ranks:
-                for gpu_map in slot_ranks['gpu_map']:
-                    rank_env[str(rank_id)] = \
-                        'export CUDA_VISIBLE_DEVICES=%s' % \
-                        ','.join([str(g) for g in gpu_map])
-                    rank_id += 1
+            for rank_id,slot in enumerate(slots):
+                rank_env[str(rank_id)] = \
+                    'export CUDA_VISIBLE_DEVICES=%s' % \
+                    ','.join([str(g.index) for g in slot['gpus']])
```

Comment:
sorry, this should be treated as a dictionary (initially I thought it is a `RO` object, but it is here after serialization) -> `g['index']`

Reply:
Yep